File: src/api/project.ts
import { invoke } from './tauri'
import type { Project, CreateProjectRequest, UpdateProjectRequest, Skill } from '../types'

export const projectApi = {
  list: async (): Promise<Project[]> => {
    return await invoke('get_projects')
  },
  
  create: async (request: CreateProjectRequest): Promise<Project> => {
    return await invoke('create_project', { request })
  },
  
  update: async (id: string, request: UpdateProjectRequest): Promise<Project> => {
    return await invoke('update_project', { id, request })
  },
  
  delete: async (id: string): Promise<void> => {
    return await invoke('delete_project', { id })
  },
}

export const projectSkillApi = {
  list: async (projectId: string, presetId?: string): Promise<Skill[]> => {
    return await invoke('get_project_skills', { projectId, presetId })
  },

  listByPresets: async (projectId: string): Promise<Record<string, Skill[]>> => {
    const result = await invoke<[string, Skill[]][]>('get_project_skills_by_presets', { projectId })
    return Object.fromEntries(result)
  },

  remove: async (projectId: string, skillId: string): Promise<void> => {
    return await invoke('remove_project_skill', { projectId, skillId })
  },

  toggleStatus: async (projectId: string, skillId: string): Promise<void> => {
    return await invoke('toggle_project_skill_status', { projectId, skillId })
  },

  batchRemove: async (projectId: string, skillIds: string[]): Promise<void> => {
    return await invoke('batch_remove_project_skills', { projectId, skillIds })
  },

  batchToggleStatus: async (projectId: string, skillIds: string[]): Promise<void> => {
    return await invoke('batch_toggle_project_skills_status', { projectId, skillIds })
  },

  checkExists: async (projectId: string, presetId: string, skillName: string): Promise<boolean> => {
    return await invoke('check_project_skill_exists', { projectId, presetId, skillName })
  },
}

File: src-tauri/src/commands/project.rs
use tauri::State;

use crate::db::connection::get_connection;
use crate::db::connection::DbConnection;
use crate::models::project::{CreateProjectRequest, Project, UpdateProjectRequest};
use crate::services::project_service;

#[tauri::command]
pub fn get_projects(db: State<'_, DbConnection>) -> Result<Vec<Project>, String> {
    let conn = get_connection(&db).map_err(|e| e.to_string())?;
    project_service::get_projects(&conn).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn create_project(
    db: State<'_, DbConnection>,
    request: CreateProjectRequest,
) -> Result<Project, String> {
    let conn = get_connection(&db).map_err(|e| e.to_string())?;
    project_service::create_project(&conn, request).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn delete_project(db: State<'_, DbConnection>, id: String) -> Result<(), String> {
    let conn = get_connection(&db).map_err(|e| e.to_string())?;
    project_service::delete_project(&conn, &id).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn update_project(
    db: State<'_, DbConnection>,
    id: String,
    request: UpdateProjectRequest,
) -> Result<Project, String> {
    let conn = get_connection(&db).map_err(|e| e.to_string())?;
    project_service::update_project(&conn, &id, request).map_err(|e| e.to_string())
}
